src/GraphBLAS-sharp/Implementations.fs

```diff
@@ -112,31 +112,1587 @@ type CSRMatrix<'a when 'a : struct and 'a : equality>(csrTuples: CSRFormat<'a>)
+
+    let mutable rows, columns, values = elements |> Array.unzip3
+    member this.Rows with get() = rows
+    member this.Columns with get() = columns
+    member this.Values with get() = values
+    member this.Elements with get() = elements
+
+    override this.Mask = Some <| Mask2D(Array.zip this.Rows this.Columns, this.RowCount, this.ColumnCount, false)
+    override this.Complemented = Some <| Mask2D(Array.zip this.Rows this.Columns, this.RowCount, this.ColumnCount, true)
+
+    member private this.SetCOOMatrix
+        (mask: Mask2D)
+        (matrix: COOMatrix<'a>) =
+        let newElements = matrix.Elements |> Array.filter (fun (i, j, _) -> mask.[i, j])
+        let elementsToSet, elementsToAppend =
+            newElements
+            |> Array.partition (
+                fun (i, j, _) ->
+                this.Elements
+                |> Array.exists (fun (u, v, _) -> (u, v) = (i, j)))
+
+        let workflow =
+            opencl {
+                let command =
+                    <@
+                        fun (ndRange: _1D)
+                            (resultValuesBuffer: 'a[])
+                            (resultRowsBuffer: int[])
+                            (resultColumnsBuffer: int[])
+                            (valuesToSetBuffer: 'a[])
+                            (rowsToSetBuffer: int[])
+                            (columnsToSetBuffer: int[]) ->
+
+                            let i = ndRange.GlobalID0
+                            let value, row, column = valuesToSetBuffer.[i], rowsToSetBuffer.[i], columnsToSetBuffer.[i]
+                            for j in 0 .. resultValuesBuffer.Length - 1 do
+                                if resultRowsBuffer.[j] = row && resultColumnsBuffer.[j] = column then
+                                    resultValuesBuffer.[j] <- value
+                    @>
+
+                let resultValues = this.Values
+                let resultRows = this.Rows
+                let resultColumns = this.Columns
+                let rowsToSet, columnsToSet, valuesToSet = elementsToSet |> Array.unzip3
+                let binder kernelP =
+                    let ndRange = _1D(elementsToSet.Length)
+                    kernelP
+                        ndRange
+                        resultValues
+                        resultRows
+                        resultColumns
+                        valuesToSet
+                        rowsToSet
+                        columnsToSet
+
+                do! RunCommand command binder
+                return! ToHost resultValues
+            }
+
+        let renewedValues = oclContext.RunSync workflow
+
+        let resultRows, resultColumns, resultValues =
+            (rows, columns, renewedValues)
+            |||> Array.zip3
+            |> Array.append elementsToAppend
+            |> Array.unzip3
+
+        rows <- resultRows
+        columns <- resultColumns
+        values <- resultValues
+
+    override this.Item
```

Comment:
I think, it's a good idea to implement all functionality of methods inside private members (or let bindings) and then call them in public members. This will make it easier to find the implementation of a specific method inside the class.